common/io/src/buffer/nested_checkpoint_reader.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io;
+use std::io::Result;
+
+use super::BufferRead;
+
+// this struct can avoid NestedCheckpointReader<NestedCheckpointReader<R>> in recursive functions
+pub struct NestedCheckpointReader<R: BufferRead> {
```

Comment:
Is that`CheckpointReader`  renaming to `NestedCheckpointReader` ?

Reply:
no,  it can maintain multi checkpoints.  diff implementation

keep using the buffer instead of preadd to the buffer of the inner reader when rollback